src/nvim/event/socket.c

```diff
@@ -81,23 +57,49 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
   watcher->cb = NULL;
   watcher->close_cb = NULL;
   watcher->events = NULL;
+
+  xfree(addr);
 }
 
 int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
   FUNC_ATTR_NONNULL_ALL
 {
   watcher->cb = cb;
-  int result;
+  int result = -EINVAL;
 
   if (watcher->stream->type == UV_TCP) {
-    result = uv_tcp_bind(&watcher->uv.tcp.handle,
-                         (const struct sockaddr *)&watcher->uv.tcp.addr, 0);
+    struct addrinfo *ai = watcher->uv.tcp.addrinfo;
+
+    for (; ai; ai = ai->ai_next) {
+      struct sockaddr_storage sas = *(struct sockaddr_storage *)ai->ai_addr;
```

Comment:
Just put `struct sockaddr_storage sas;` before `uv_tcp_getsockname()`. We get the socket name from the socket descriptor, not from the content of `sas`.